### PR TITLE
Agent57: Outperforming the human Atari benchmark

### PR DESCRIPTION
**Pull request by AntoineTheb**

### CODE REVIEW

File: article/_posts/2020-06-05-Agent57.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+layout: review
+title: "Agent57: Outperforming the human Atari benchmark"
+tags: reinforcement deep-learning RNN
+author: "Antoine Théberge"
+cite:
+    authors: "Adrià Puigdomènech Badia, Bilal Piot, Steven Kapturowski, Pablo Sprechmann, Alex Vitvitskyi, Daniel Guo, Charles Blundell"
+    title:   "Agent57: Outperforming the Atari Human Benchmark"
+    venue:   "Arxiv"
+pdf: ""
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:10]: Missing link to the arXiv PDF.

File: article/_posts/2020-06-05-Agent57.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+# Highlights
+
+- Agent57 is able to perform well on all 57 Atari games in the ALE[^1]
+- Q-function is split in two to decompose the contribution of intrinsic and extrinsic rewards
+- A meta-controller modeled as a bandit problem controls the tradeoff between exploration and exploitation
+
+# Introduction
+
+![](/article/images/Agent57/improvements.svg)
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:22]: The picture looks huge when displayed. Was this intentional?

File: article/_posts/2020-06-05-Agent57.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    authors: "Adrià Puigdomènech Badia, Bilal Piot, Steven Kapturowski, Pablo Sprechmann, Alex Vitvitskyi, Daniel Guo, Charles Blundell"
+    title:   "Agent57: Outperforming the Atari Human Benchmark"
+    venue:   "Arxiv"
+pdf: ""
+---
+
+
+# Highlights
+
+- Agent57 is able to perform well on all 57 Atari games in the ALE[^1]
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:16]: Define what `ALE` is on this first appearance?

File: article/_posts/2020-06-05-Agent57.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+- Q-function is split in two to decompose the contribution of intrinsic and extrinsic rewards
+- A meta-controller modeled as a bandit problem controls the tradeoff between exploration and exploitation
+
+# Introduction
+
+![](/article/images/Agent57/improvements.svg)
+A condensed history of improvements that led to Agent57
+> source: Deepmind blog
+
+NGU[^2] was a monumental success as it was the first agent to achieve good performance on hard-exploration games. However, it conversely failed to gain good scores on "easier" games. The authors hypothesize that this is because, at each episode, NGU uniformly assign a family of more explorative or exploitative agents at random to gather episodes. This process, according to the authors, is suboptimal as the need for exploration or exploitation varies when learning a game as well as with different games.
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:26]: Define what `NGU` is on this first appearance?

File: article/_posts/2020-06-05-Agent57.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+where $$j$$ is a one-hot vector pointing to the $$(\beta, \gamma)$$ pair controlling exploration and discount for the actor, just like in NGU. The authors show in the appendix that this decomposition is valid and that the optimization of this Q-function is equivalent to the optimization of a single Q-function with reward $$r^e + \beta_j r^i$$.
+
+## Adaptive Exploration over a Family of Policies
+
+NGU uses a family of explorative and exploitative policies during training and trains them all equally, regardless of their actual contributions to the learning process. The authors propose using a meta-controller to select which policies will be used at training and testing time. To do so, the authors propose posing the policy selection as a non-stationary multi-arm bandit problem. At the beginning of each episode, the meta controller selects a bandit "arm" to assign a pair $$(\beta, \gamma)$$ per actor. 
+
+Because the needs for exploration and exploitation change over time, the bandit problem is considered non-stationary. This means that an algorithm like UCB[^3] would not work. To alleviate the problem, the authors employ a "sliding-window" UCB, which is described in the appendix.
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:42]: Define what `UCB` is on this first appearance?

File: article/_posts/2020-06-05-Agent57.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+The authors propose to split the state-action value function (Q) into this:
+
+$$Q(x, a, j; \theta) = Q(x,a,j;\theta^e) + \beta_j Q(x,a,h;\theta^i)$$
+
+where $$j$$ is a one-hot vector pointing to the $$(\beta, \gamma)$$ pair controlling exploration and discount for the actor, just like in NGU. The authors show in the appendix that this decomposition is valid and that the optimization of this Q-function is equivalent to the optimization of a single Q-function with reward $$r^e + \beta_j r^i$$.
+
+## Adaptive Exploration over a Family of Policies
+
+NGU uses a family of explorative and exploitative policies during training and trains them all equally, regardless of their actual contributions to the learning process. The authors propose using a meta-controller to select which policies will be used at training and testing time. To do so, the authors propose posing the policy selection as a non-stationary multi-arm bandit problem. At the beginning of each episode, the meta controller selects a bandit "arm" to assign a pair $$(\beta, \gamma)$$ per actor.
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:40]: `(...) which policies will be used at training and testing time.`

Maybe I'm missing something, but does this imply that at test time a *different* policy is used wrt to the one used at train time? I'd assume that the same policy will be used in both, but which policy is used is defined/learned by the meta-controller during training. Am I wrong?

File: article/_posts/2020-06-05-Agent57.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+## Adaptive Exploration over a Family of Policies
+
+NGU uses a family of explorative and exploitative policies during training and trains them all equally, regardless of their actual contributions to the learning process. The authors propose using a meta-controller to select which policies will be used at training and testing time. To do so, the authors propose posing the policy selection as a non-stationary multi-arm bandit problem. At the beginning of each episode, the meta controller selects a bandit "arm" to assign a pair $$(\beta, \gamma)$$ per actor. 
+
+Because the needs for exploration and exploitation change over time, the bandit problem is considered non-stationary. This means that an algorithm like UCB[^3] would not work. To alleviate the problem, the authors employ a "sliding-window" UCB, which is described in the appendix.
+
+## Data
+
+The agent was trained using a single set of hyperparameters on all 57 games in the ALE[^1] (separately)
```
Comment (jhlegarreta) [inline article/_posts/2020-06-05-Agent57.md:46]: `(...) single set of hyperparameters on all 57 games in the ALE[^1] (separately)`

Seems unclear to me: were the agents trained separately using the same set of hyper-parameters?

File: article/_posts/2020-06-05-Agent57.md (hunk 2 above)
Comment (AntoineTheb) [reply article/_posts/2020-06-05-Agent57.md:22]: Kinda, I have scaled it down a bit.

File: article/_posts/2020-06-05-Agent57.md (hunk 6 above)
Comment (AntoineTheb) [reply article/_posts/2020-06-05-Agent57.md:40]: The meta-controller also does select which policy will be used at test time. This allows the agent to produce different policies at test time and therefore different behaviors can be had. I'll reword this part to make this clearer.

File: article/_posts/2020-06-05-Agent57.md (hunk 7 above)
Comment (AntoineTheb) [reply article/_posts/2020-06-05-Agent57.md]: Indeed. I will make this part clearer

**Comment by AntoineTheb** [comment]
Thank you for the comments @jhlegarreta, I have addressed them in 0460679

**Comment by jhlegarreta** [comment]
> Thank you for the comments @jhlegarreta, I have addressed them in 0460679

:+1: .

**Pull request merged**